deepspeed/checkpoint/ds_to_universal.py

```diff
@@ -0,0 +1,298 @@
+            if pp_index > 0 and any(re.match(pattern, name) for pattern in pipeline_replicated_params):
+                # Skip tied weights that are replicated in first and last pp stages
+                continue
+
+            # pprint(f"dpt{dp_index}{pp_index}{tp_index} {param_group_id} {name} => {fragment_mapping.start}:{fragment_mapping.numel}")
+            for state_key in flat_state.keys():
+                dump_param_fragment(dir, tp_index, dp_index, state_key, flat_state[state_key], name,
+                                    fragment_mapping.start, fragment_mapping.numel)
+
+
+cnt = 0
+
+
+def dump_param_fragment(dir, tp_index, dp_index, state_name, state_flat_tensor, param_name, offset, numel):
+
+    global cnt  # temp hack
+
+    param_base_path = os.path.join(dir, param_name, str(tp_index))
+    os.makedirs(param_base_path, exist_ok=True)
+
+    cnt += 1
+    counter = f"{dp_index:0>2d}"
+
+    path = os.path.join(param_base_path, f"{state_name}.{counter}")
+
+    #print(f"{param_name}: {offset}: {numel} => {path}")
+
+    t = state_flat_tensor.narrow(0, offset, numel)
+    _save_checkpoint(path, t)
+
+
+def _merge_zero_shards(param_base_path, state, tp_degree, slice_shape):
+    slices = []
+    for tp_index in range(tp_degree):
+        prefix_path = os.path.join(param_base_path, str(tp_index), f"{state}")
+        paths = sorted(list(glob.glob(f"{prefix_path}.0*")))
+        shards = [torch.load(p) for p in paths]
+        slice = torch.cat(shards, dim=0).reshape(slice_shape)
+        slices.append(slice)
+
+    return slices
+
+
+def _get_vocab_divisibility_padding_tensor(ds_checkpoint, padded_vocab_tensor):
+    checkpoint_info = ds_checkpoint.get_checkpoint_info()
+    if padded_vocab_tensor.shape[0] > checkpoint_info[ORIGINAL_VOCAB_SIZE]:
+        return padded_vocab_tensor[-1]
+    else:
+        return torch.zeros(padded_vocab_tensor.shape[1])
+
+
+def merge_tp_slices(ds_checkpoint, dir, slice_dir, tp_degree, name_and_shape):
+    name, shape = name_and_shape
+    slice_base_path = os.path.join(slice_dir, name)
+    param_base_path = os.path.join(dir, name)
+
+    universal_checkpoint_info = ds_checkpoint.get_checkpoint_info(UNIVERSAL_CHECKPOINT_INFO)
+    parameters_to_average = universal_checkpoint_info.get(PARAMETER_TO_AVERAGE_PATTERNS, [])
+    parameters_with_row_parallelism = universal_checkpoint_info.get(PARAMETER_WITH_ROW_PARALLELISM_PATTERNS, [])
+    vocabulary_parameters = universal_checkpoint_info.get(VOCABULARY_PARAMETER_PATTERNS, [])
+    for state in ("fp32", "exp_avg", "exp_avg_sq"):
+        slices = _merge_zero_shards(slice_base_path, state, tp_degree, shape)
+        final_path = os.path.join(param_base_path, f"{state}.pt")
+
+        #print(f"Expected shape: {shape}")
+        #print(f"Fragment sizes:", list(frag.shape for frag in slices))
+        ckpt_dict = {}
+        if any(re.match(pattern, name) for pattern in parameters_to_average):
+            param = sum(slices) / len(slices)
```

Comment:
parameters_to_average are expected to be the same.
therefore, instead of doing average, suggest to:
1. assert all are equal
2. return first one

Reply:
Good point. I may be wrong, but I seem to recall that there was a case of averaging across the tensor slices in BLOOM, separately from the case of replicated slices. Regardless, it makes sense to separate these two cases.

@stas00, any thoughts?

Reply:
It has been a very long time, and I unfortunately don't remember this exact use case, but I'm pretty sure we had to average some of the tensors. Otherwise we would have copied the first one as it simpler and clearly isn't the same as averaging.

Probably the safest approach is to give the choice to the user: so that they could define which params are to be `parameters_copy_rank0` and which are `parameters_to_average`. And assert for the former if it's not equal on all ranks to flag to the user that they might be doing the wrong thing.

Reply:
oh, but wait, this is not Megatron-Deepspeed - this is ZeRO - so the logic is quite different. I thought you were asking about Meg-DS as a similar PR went there as well.

Reply:
@stas00, this is indeed for Meg-DS. I have moved `ds_to_universal.py` into DS codebase for future support, as we discussed.   So, this is that general problem of how to reshape partitioned tensor slices.

Reply:
Aha, that explains it, thank you.

So this is the problem of whether to average say TP tensor of rank 0 across all replicas, yes? i.e. to average or not is across replicas.

They should be the same, except when they aren't. And I think we saw in the case of BLOOM that they weren't the same, which I think was a bug. But again, this was long time ago to remember the details.

That's why I suggest to give the reins to the user by defining both approaches and choosing a sensible default.

Reply:
Done. Now, users can specify whether slices are replicated or averaged.